cmd/fleetctl/mdm.go

```diff
@@ -52,55 +53,108 @@ func mdmRunCommand() *cli.Command {
-			if err := client.CheckMDMEnabled(); err != nil {
+			if err := client.CheckAnyMDMEnabled(); err != nil {
 				return err
 			}
 
-			hostIdent := c.String("host")
+			// dedupe and remove any empty host identifier
+			hostIdents := c.StringSlice("hosts")
+			slices.Sort(hostIdents)
+			hostIdents = slices.Compact(hostIdents)
+			if len(hostIdents) > 0 && hostIdents[0] == "" {
+				// because it is sorted, an empty ident can only be at the start
+				hostIdents = hostIdents[1:]
+			}
+			if len(hostIdents) == 0 {
+				return errors.New(`Required flag "hosts" not set`)
+			}
+
 			payloadFile := c.String("payload")
 			payload, err := os.ReadFile(payloadFile)
 			if err != nil {
 				return fmt.Errorf("read payload: %w", err)
 			}
 
-			host, err := client.HostByIdentifier(hostIdent)
-			if err != nil {
-				var nfe service.NotFoundErr
-				if errors.As(err, &nfe) {
-					return errors.New("The host doesn't exist. Please provide a valid hostname, uuid, osquery_host_id or node_key.")
-				}
-				var sce kithttp.StatusCoder
-				if errors.As(err, &sce) {
-					if sce.StatusCode() == http.StatusForbidden {
-						return fmt.Errorf("Permission denied. You don't have permission to run an MDM command on this host: %w", err)
+			// fetch all specified hosts by their identifier
+			//
+			// Note that this retrieves all hosts sequentially as it assumes the
+			// number of hosts will typically be small. If we need to improve on
+			// this, we could use something like the errgroup package to retrieve
+			// them concurrently or better yet add a batch endpoint to get hosts by
+			// the loose "identifier".
+			var (
+				hostUUIDs     []string
+				notFoundCount int
+				platform      string
+			)
+			for _, ident := range hostIdents {
+				host, err := client.HostByIdentifier(ident)
+				if err != nil {
+					var nfe service.NotFoundErr
+					if errors.As(err, &nfe) {
+						notFoundCount++
+						continue
 					}
+
+					var sce kithttp.StatusCoder
+					if errors.As(err, &sce) {
+						if sce.StatusCode() == http.StatusForbidden {
+							return fmt.Errorf("You don't have permission to run an MDM command on one or more specified hosts: %w", err)
+						}
+					}
+					return err
 				}
-				return err
+
+				if host.Platform != platform && platform != "" {
+					return errors.New(`Command can't run on hosts with different platforms. Make sure the hosts specified in the "hosts" flag are either all macOS or all Windows hosts.`)
+				}
+				platform = host.Platform
+
+				// TODO(mna): this "On" check is brittle, but looks like it's the only
+				// enrollment indication we have right now...
+				if host.MDM.EnrollmentStatus == nil || !strings.HasPrefix(*host.MDM.EnrollmentStatus, "On") ||
+					host.MDM.Name != fleet.WellKnownMDMFleet {
+					// TODO(mna): on my manual QA with a Windows VM, the mdm name was
+					// FleetDM, not Fleet. If that is correct and expected, we should
+					// change the WellKnown constant (or accept both)?
```

Comment:
Anyone knows if this is a bug that needs addressing or if I've done something wrong in my VM enrollment/QA? I have vague memories of hearing a discussion about that recently-ish?

Reply:
@mna you did nothing wrong! we fixed this in `main`. 

https://github.com/fleetdm/fleet/blob/8efac8b648c2f7de0694b36ff19fb556b8336eb5/server/mdm/microsoft/microsoft_mdm.go#L181

should we update the branch?

Reply:
Ah nice thank you! I guess it's fine for now, unless the merge of `main` into the feature-branch is straightforward. As long as QA is done after we merge to `main` it should be fine, it's just something to be aware of if folks do some manual QA on the feature branch between now and the merge to main (in my case I simply updated the name in the DB).

Reply:
@roperzh So I just copied over the fix that you linked, just to avoid any potential debugging/frustration while testing on the feature branch. Hopefully adjusting the constant is all that was necessary!

Reply:
@mna IIRC that was it! I was about to mention the tests, I think you're just missing this one:

https://github.com/fleetdm/fleet/blob/3e815479f906773c8538a0aa798ce43a5fb6d877/server/service/mdm_test.go#L256